src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
+                             $ TypeError { _typeErrorExpected = "list"
+                                         , _typeErrorActual   = "integer"
+                                         }) e k
+         )
+      (\err -> Er err env kont)
+
+
+    -- lists
+    -- base case
+    (InList scope [] dones env kont) ->
+      evalAsSyntax v
+      (\good -> Down (unCore scope) e (InScope (List $ reverse $ good : dones) env kont))
+      (\err  -> Er err env kont)
+    -- still some todo
+    (InList scope (el:els) dones env kont) ->
+      evalAsSyntax v
+      (\good -> Down (unCore el) env (InList scope els (good : dones) env kont))
+      (\err  -> Er err env kont)
+
+
+    -- Macros
+    (InPureMacro env kont) -> Up (ValueMacroAction $ MacroActionPure v) env kont
+    (InBindMacroHd tl env kont) ->
+      evalAsMacroAction v
+      (\good -> Down (unCore tl) env (InBindMacroTl good env kont))
+      (\err  -> Er err env kont)
+
+    (InBindMacroTl macroAction env kont) ->
+      evalAsClosure v
+      (\good -> Up (ValueMacroAction $ MacroActionBind macroAction good) env kont)
+      (\err  -> Er err env kont)
+
+
+    -- Syntax and Atomics
+    (InInteger scope env kont) ->
+      evalAsInteger v
+      (\good -> Down (unCore scope) env (InScope (Integer good) env kont))
+      (\err  -> Er err env kont)
+    (InString scope env kont) ->
+      evalAsString v
+      (\good -> Down (unCore scope) env (InScope (String good) env kont))
+      (\err  -> Er err env kont)
+    (InLoc stx env kont) ->
+      evalAsSyntax v
+      (\(Syntax (Stx _ newLoc _)) -> Down (unCore stx) env (InLocStx newLoc env kont) )
+      (\err -> Er err env kont)
+    (InLocStx loc env kont) ->
+      evalAsSyntax v
+      (\(Syntax (Stx scs _ contents)) -> Up (ValueSyntax $ Syntax $ Stx scs loc contents) env kont)
+      (\err -> Er err env kont)
+    (InLog   env kont)   ->
+      evalAsSyntax v
+      (\good -> Up (ValueMacroAction (MacroActionLog good)) env kont)
+      (\err  -> Er err env kont)
+
+
+    -- Errors
+    (InError env kont) ->
+      evalAsSyntax v
+      (\good -> Er (EvalErrorUser good) env kont)
+      (\err  -> Er err env kont)
+    (InSyntaxErrorMessage locs env kont) ->
+      evalAsSyntax v
+      (\msg_syn ->
+         case locs of
+           -- done
+           []     -> Up (ValueMacroAction $ MacroActionSyntaxError
+                          (SyntaxError { _syntaxErrorMessage   = msg_syn
+                                       , _syntaxErrorLocations = mempty
+                                       })) env kont
+           (l:ls) -> Down (unCore l) env (InSyntaxErrorLocations msg_syn ls mempty env kont)
+          )
+      (\err -> Er err env kont)
+    -- done
+    (InSyntaxErrorLocations msg_syn [] dones env kont) ->
+        Up (ValueMacroAction
+                $ MacroActionSyntaxError (SyntaxError { _syntaxErrorMessage   = msg_syn
+                                                      , _syntaxErrorLocations = dones
```

Comment:
```suggestion
                                                      , _syntaxErrorLocations = reverse (v : dones)
```

Reply:
I believe this is a type error:

- `v : Value`
- `dones : [Syntax]`

here is what I believe the original case is in `eval`:

```
eval (Core (CoreSyntaxError syntaxErrorExpr)) = do
  syntaxErrorValue <- traverse evalAsSyntax syntaxErrorExpr
  pure $ ValueMacroAction
       $ MacroActionSyntaxError syntaxErrorValue
```

is there an example that exhibits this bug?

Reply:
> I believe this is a type error: [...]

You're right, a call to `evalAsSyntax` is also needed.


> is there an example that exhibits this bug?

Here is one:
```
#lang kernel
(import (shift kernel 1))

(define-macros
  ([m (lambda (stx)
        (syntax-error 'error-message
                        'location-1  -- line 7
                        'location-2  -- line 8
                        'location-3  -- line 9
                        'location-4  -- line 10
                        ))]))

(example (m))
```

on main, this program outputs
```
toy.kl:7.26-7.36:
    Syntax error from macro:
    #[toy.kl:6.24-6.37]<error-message>Additional locations:
    toy.kl:8.26-8.36
    toy.kl:9.26-9.36
    toy.kl:10.26-10.36
```
whereas on this branch, this program outputs one fewer location, and outputs them in the opposite order:
```
toy.kl:9.26-9.36:
    Syntax error from macro:
    #[toy.kl:6.24-6.37]<error-message>Additional locations:
    toy.kl:8.26-8.36
    toy.kl:7.26-7.36
```
The first location is distinguished because it is printed before the error message, so let's use the shorthand `[7], 8, 9, 10` to describe the output on main and `[9], 8, 7` to describe the output on this branch.

Line 10 (`'location-4`) is missing, because `step (Up v e k)` is supposed to insert the value `v` into the context `k`, but here when the context is `InSyntaxErrorLocations`, the `v` is not inserted, it is dropped on the floor.

And the locations are listed in the opposite order because as you traverse `InSyntaxErrorLocations`'s `[Core]` from left to right, you prepend each output to its `[Syntax]`, so the right-most `Core` ends up being the leftmost `Syntax`. You have fixed this elsewhere by calling `reverse` in the `InCtor []` and `InList []` cases, you just forgot to also do it in the `InSyntaxErrorLocations  []` case.